### ISSUE TITLE
Tensorboard port utilization

### ISSUE THREAD

**Issue by gogasca** [linkedin/TonY#220]
**Setup**
Currently I have a Dataproc cluster of 4 instances. (1 master + 4 NodeManagers).

I have available in total resources for cluster: 48 GB RAM 16 vCPU

![image](https://user-images.githubusercontent.com/30065079/54875139-f7017d00-4db5-11e9-82d8-b75b0f27ae97.png)

When I launch a 3 TensorFlow jobs:
Each TensorFlow job: (12 GB RAM 4 vCPU):
- 1 TF Parameter Server 2GB RAM 1 vCPU
- 1 TF Worker 4GB RAM 1 vCPU
- 1 TF Worker 4GB RAM 1 vCPU
- 1 application container 2 GB 1 vCPU

Some workers share the same NodeManager, hence TensorBoard gets started more than once on each Node.

Due to configuration in Python mnist_distributed.py
port = os.getenv('TB_PORT_ENV_VAR', 6006)

We read from Environment variable, and start TB on this port.

**Version**
Hadoop 2.9

When multiple jobs start, binding fails:
Shouldn't TonY validate the port before setting up the value in Environment variable?

```
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "mnist_distributed.py", line 260, in <module>
    tf.app.run()
  File "/mnt/sdb/hadoop/yarn/nm-local-dir/usercache/root/appcache/application_1553318393788_0009/container_1553318393788_0009_01_000003/venv/tf/lib/python3.5/site-packages/tensorflow/
python/platform/app.py", line 125, in run
    _sys.exit(main(argv))
  File "mnist_distributed.py", line 216, in main
    start_tensorboard(FLAGS.working_dir)
  File "mnist_distributed.py", line 184, in start_tensorboard
    tb.launch()
  File "/mnt/sdb/hadoop/yarn/nm-local-dir/usercache/root/appcache/application_1553318393788_0009/container_1553318393788_0009_01_000003/venv/tf/lib/python3.5/site-packages/tensorboard
/program.py", line 250, in launch
    server = self._make_server()
  File "/mnt/sdb/hadoop/yarn/nm-local-dir/usercache/root/appcache/application_1553318393788_0009/container_1553318393788_0009_01_000003/venv/tf/lib/python3.5/site-packages/tensorboard
/program.py", line 306, in _make_server
    return self.server_class(app, self.flags)
  File "/mnt/sdb/hadoop/yarn/nm-local-dir/usercache/root/appcache/application_1553318393788_0009/container_1553318393788_0009_01_000003/venv/tf/lib/python3.5/site-packages/tensorboard
/program.py", line 407, in __init__
    port)
tensorboard.program.TensorBoardServerException: TensorBoard could not bind to port 6006, it was already in use
```

**Workaround:**

```
import socket
tf.flags.DEFINE_integer("max_tb_retries", 3, "The Max number of times we attempt to start TB")

def get_free_port(port):
    retries = 0
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)    
    while retries <= FLAGS.max_tb_retries:
        try:
            sock.bind(('', port))
            sock.close()
            return port
        except OSError:
            port += 1
            retries += 1
    raise IOError('TensorBoard port in use')

def start_tensorboard(logdir):
    tb = tb_program.TensorBoard(plugins=[core_plugin.CorePluginLoader()])
    port = get_free_port(os.getenv('TB_PORT_ENV_VAR', 6006))
    tb.configure(logdir=logdir, port=port)
    tb.launch()
    logging.info("Starting TensorBoard with --logdir=%s" % logdir)
```

**Logs**
[application_1553318393788_0009.log](https://github.com/linkedin/TonY/files/3000021/application_1553318393788_0009.log)

**Comment by oliverhu** [comment]
it should be **TB_PORT** not "TB_PORT_ENV_VAR".. @gogasca  could you update the example?

**Comment by oliverhu** [comment]
fixed in #221
